IMouse / Mouse refactoring

dpisanu opened the pull request:
Cleaning up the Mouse Classes and Interfaces
This 'should' not break current implementation. 
It's just separating some of the code into manageable chunks.

BareMetalMouse
+ Added new Class called BareMetalMouse
+ Contains the DLLImport from user32
+ Contains the basic Button Interactions

IBaseMouse / BaseMouse
+ New Interface called IBaseMouse
+ New Abstract Class called BaseMouse
+ Moved most of the Functions from Mouse to BaseMouse
+ Interface now reflects all the public Functions and Properties from
the Class
+ BaseMouse contains all interactions that do not directly require an
ActionPerformed on the IActionListener

IMouse / Mouse
+ Cleaned up Mouse -> Move the Bare Mouse work to BareMetalMouse
+ Cleaned up Mouse -> Moved general Mouse Work to BaseMouse
+ Mouse now only contains the code for Mouse Interaction sporting an
IActionListener
+ Interface reflects the public Functions and properties

AttachedMouse
+ Inherits from IBaseMouse now
+ Offering more Function Calls now
+ On Operations where Mouse offers a version with IActionListener, that
Function is called
+ On Operations where Mouse does not offer an IActionListener call,, the
ActionPerformed Function from Mouse is called

dpisanu commented:
I would like to pull these changes through White that not the concrete Mouse Type is being used but IMouse.
Idea is to always just go with the Interface and not the concrete implementation as Parameter.

Roemer commented:
There seems to be at least one virtual missing ;) I will review this PR later. But from the log I like it. I saw the same with the Configuration, so if you want you can clean this up as well :)

Roemer commented on src/TestStack.White/InputDevices/BareMetalMouse.cs:
```diff
@@ -0,0 +1,157 @@
+            MouseButtonDown(MouseButton.Left);
+        }
+
+        public static void RightUp()
```
I think those static methods should be removed or at least be marked as obsolete.

Roemer commented on src/TestStack.White/InputDevices/BareMetalMouse.cs:
```diff
@@ -0,0 +1,157 @@
+
+namespace TestStack.White.InputDevices
+{
+    public class BareMetalMouse
```
Shouldn't we mark this class as static since it only contains static methods?

Roemer commented on src/TestStack.White/InputDevices/BareMetalMouse.cs:
```diff
@@ -0,0 +1,157 @@
+
+        #region Public
+
+        public static void LeftUp()
```
I think those 5 methods should be removed or marked as obsolete.

Roemer commented on src/TestStack.White/InputDevices/AttachedMouse.cs:
```diff
@@ -8,62 +8,169 @@ namespace TestStack.White.InputDevices
     /// performed the window, from which it was retreived, is brought to focus if it is not.
     /// </summary>
-    public class AttachedMouse : IMouse
+    public class AttachedMouse : IBaseMouse
```
I still don't know what this class is for. Can't we remove it?

Roemer commented on src/TestStack.White/InputDevices/IBaseMouse.cs:
```diff
@@ -0,0 +1,204 @@
+        #region Mouse Click
+
+        /// <summary>
+        /// Perform a Click Operation
```
Note that it's a left click

Roemer commented on src/TestStack.White/InputDevices/IBaseMouse.cs:
```diff
@@ -0,0 +1,204 @@
+        /// Perform a Click Operation
+        /// </summary>
+        /// <param name="point">Point where to Click on</param>
+        void Click(Point point);
```
Note that it's a left click

Roemer commented on src/TestStack.White/InputDevices/IBaseMouse.cs:
```diff
@@ -0,0 +1,204 @@
+        /// Perform a Double Click Operation
+        /// </summary>
+        /// <param name="point">Point where to Double Click on</param>
+        void DoubleClick(Point point);
```
Note that it's a left click

Roemer commented on src/TestStack.White/InputDevices/IBaseMouse.cs:
```diff
@@ -0,0 +1,204 @@
+
+        #endregion
+
+        #region Mouse Click
```
Maybe add a separate region for left click

dpisanu commented:
The AttachedMouse contains it's IActionListener that it tries to send the Action on.
Where as the Mouse is IActionListener less.

It basecally wraps Mouse and handles the Action sending to the Listener so you do not have to send it yourself everytime you call an function.

Same thing for the Keyboard. I am working on that one right now. Will be a separate Request.
I'll get on your Remarks once I am done.

dpisanu commented:
Ok. Fixed the Virtuals in AttachedMouse
Added Click and DoubleClick in BaseMouse
Added LeftClick and LeftDoubleClick in BaseMouse, Mouse and AttachedMouse
Marked the Click and DoubleClick as Obsolete.
Adapted all White internal calls to the Obsolete Functions.
Now, anybody that uses White has Obsolete marked as an Error during compile, integration of White will break. If it's just treated as a Warning, nothing will break.

dpisanu commented:
Just cleaned up the statics and made the class static as well.

Roemer commented on src/TestStack.White.sln:
```diff
@@ -1,7 +1,7 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 2013
-VisualStudioVersion = 12.0.40629.0
+VisualStudioVersion = 12.0.31101.0
```
Seems like you're on Visual Studio Update 4 instead of 5. Either update your VS or don't include this change

Roemer commented on src/TestStack.White/InputDevices/Mouse.cs:
```diff
@@ -1,145 +1,32 @@
-        static extern int GetSystemMetrics(SystemMetric smIndex);
+        #region Fields
 
         public static Mouse Instance = new Mouse();
```
We should get rid of the "Instance" as well since the interface should be used. But that could lead to a bigger refactoring. Quickly check if it is easily possible and if not, it's a todo for another PR.

dpisanu commented:
sorry bout the SLN change. that slipped my click on both cases.
Those DLLIMports are actually being used. So we can't just remove them.

dpisanu commented:
So is this change ok to go on master? I would then create a separate Changeset for the UIACom Branch

dpisanu commented:
Please close this PR
